modules/gmake2/_preload.lua

```diff
@@ -62,7 +62,12 @@
 
 		onCleanProject = function(prj)
 			p.clean.file(prj, p.modules.gmake2.getmakefilename(prj, true))
-		end
+		end,
+
+		pathVars = {
+			["rule.inputs"]	 = { absolute = false, token = "$<" },
+			["rule.outputs"] = { absolute = false, token = "$@" }
```

Comment:
> For non-msbuild environments like gmake2 or codelite, these tokens will be substituted as single input and output as there're no batching.

What would be the value without `pathVars` ? (i.e for gmake/xcode/external modules)
Does they need update?

Reply:
Those tokens are only valid in `buildcommands` in custom rule context.
For built-in modules it seems like gmake2, codelite and vs are only modules support custom rules.

However I do think it's better to have fallback for any modules. I was thinking of using `fcfg_mt` for fallback and name tokens something like `file.ruleinputs`/`file.ruleoutputs`. Does that make more sense?

Reply:
Custom rules are indeed not supported by gmake and xcode.
Some external module such as codeblocks/ninja does support it.

How is `rule.outputs` "concatenated"? with space? (and `$@` `$<` may probably be removed).
`rule.inputs` is a little more complex, Does it include `%{file.relpath}`?

Reply:
Yes, in VS modules,
`rule.inputs` is space-concatenated `file.relpath`s for the batch. (Same as item included in `.vcxproject`)
`rule.outputs` is space-concatenated `path.normalize(buildoutputs)`s for the batch. (Same as outputs defined in `.targets`)